presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java

```diff
@@ -77,26 +80,57 @@ public static PlanNodeStatsEstimate groupBy(PlanNodeStatsEstimate sourceStats, C
         }
 
         double rowsCount = 1;
+        double totalSize = 0;
         for (VariableReferenceExpression groupByVariable : groupByVariables) {
             VariableStatsEstimate symbolStatistics = sourceStats.getVariableStatistics(groupByVariable);
             int nullRow = (symbolStatistics.getNullsFraction() == 0.0) ? 0 : 1;
             rowsCount *= symbolStatistics.getDistinctValuesCount() + nullRow;
+            if (!isNaN(symbolStatistics.getAverageRowSize())) {
+                totalSize += symbolStatistics.getAverageRowSize() * rowsCount;
+            }
+        }
+        // Set estimation boundary for aggregation:
+        // If the source only return a single row, it's guaranteed the upper bound is a single row output for this agg.
+        // Aggregation node cannot return output row count more than estimated row count of source node.
+        // So if source stats is present, use the output row count from the source stats as the estimated output row
+        // count as the upper bound limit to avoid returning an unknown output potentially breaks the join search space
+        if (!isNaN(rowsCount)) {
+            result.setOutputRowCount(min(rowsCount, sourceStats.getOutputRowCount()));
```

Comment:
Can we add a configurable factor-constant to multiply here?

Reply:
sure. I will add a tunable aggregate coefficient as the multiplier. Default is 100% selectivity from the source stats.